### PR TITLE
[V3] New V3 SkaffoldRunner

### PR DESCRIPTION
**Pull request by yuwenma**
**Related**:  #5673

**Description**
1. Refactored the runner directory.
2. Create a New SkaffoldRunner for V3. Both runner.SkaffoldRunner and v3.SkaffoldRunner implements the Runner interface.
2. Build, Test, Prune and other shared methods are moved from SkaffoldRunner to several individual structs (BuildRunner, TestRunner).
3. Both SkaffoldRunner and the V3 SkaffoldRunner have the shared structs as components.
4. Unittests are updated.
5. Add placeholders (not implemented error) in v3 SkaffoldRunner for later implementation.

### CODE REVIEW

File: pkg/skaffold/runner/build_deploy.go [hunk 1]
```diff
@@ -99,57 +116,13 @@ func (r *SkaffoldRunner) Build(ctx context.Context, out io.Writer, artifacts []*
...
-func (r *SkaffoldRunner) Test(ctx context.Context, out io.Writer, artifacts []graph.Artifact) error {
-	if err := r.tester.Test(ctx, out, artifacts); err != nil {
-		return err
-	}
-
-	return nil
-}
-
-// DeployAndLog deploys a list of already built artifacts and optionally show the logs.
-func (r *SkaffoldRunner) DeployAndLog(ctx context.Context, out io.Writer, artifacts []graph.Artifact) error {
```
Comment (yuwenma) [inline pkg/skaffold/runner/build_deploy.go:112]: Moved to Deploy.go

File: pkg/skaffold/runner/v3/generate_pipeline.go [hunk 2]
```diff
@@ -0,0 +1,28 @@
+/*
```
Comment (MarlonGamez) [inline pkg/skaffold/runner/v3/generate_pipeline.go:1]: not completely sure we'll be supporting this feature in the future, might be best to just leave it out of newer versions?
wdyt @tejal29 @nkubala ?

**Review by MarlonGamez — COMMENTED** [review]
This looks good to me, I think the thing I'm confused about is the v3 naming. I understand that it's named off the schema, but is kinda confusing since skaffold's major version number is still 1

File: pkg/skaffold/runner/v3/generate_pipeline.go (hunk 2 above)
Comment (yuwenma) [reply pkg/skaffold/runner/v3/generate_pipeline.go:1]: +1

I'd vote for removing the generate-pipeline from v3, or at most p2. 
For the v2 main stream, skaffold+ tekton seems to be a nice combination. the tekton also use skaffold in their examples see [here](https://github.com/tektoncd/pipeline/blob/main/docs/tutorial.md#running-your-task)

**Comment by yuwenma** [comment]
> This looks good to me, I think the thing I'm confused about is the v3 package naming. I understand that it's named off the schema, but is kinda confusing since skaffold's major version number is still 1

yeah, when changing the schema from v1 to v2, I think we actually should bump the skaffold version to v2 at that time. Do you suggest using v2 here or re-align the schema and master version in this new skaffold render?

File: pkg/skaffold/runner/v3/generate_pipeline.go (hunk 2 above)
Comment (tejal29) [reply pkg/skaffold/runner/v3/generate_pipeline.go:1]: hmm i can see that generated yaml would change. We could re-implement what we have as P2.

We should also instrument its usage.

**Review by tejal29 — CHANGES_REQUESTED** [review]

File: pkg/skaffold/runner/build_deploy.go [hunk 3]
```diff
@@ -26,18 +26,35 @@ import (
...
 	eventV2 "github.com/GoogleContainerTools/skaffold/pkg/skaffold/event/v2"
 	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/graph"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/kubernetes"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/runner/runcontext"
 	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/schema/latest"
 	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/tag"
 	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/util"
 )
 
+type BuildRunner struct {
```
[the remark this answers is not in the document]
Comment (yuwenma) [reply pkg/skaffold/runner/build_deploy.go]: Sure. was trying to distinguish this builder with the build.Builder.

**Review by tejal29 — APPROVED** [review]

**Pull request merged**